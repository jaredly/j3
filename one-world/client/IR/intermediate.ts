// An intermediate representation

import { Nodes, Node, Style } from '../../shared/nodes';
import { ListDisplay, RenderInfo } from '../../shared/renderables';

type Format = Style;

const refStyle: Style = {
    fontWeight: 'bold',
    // type: 'text', style: { fontWeight: 'bold' }
};

export type IR =
    | { type: 'text'; text: string; style?: Style }
    | {
          type: 'vert';
          items: IR[];
          style?: Style;
          layout?: { tightFirst: number; pairs?: number[]; indent: number };
      }
    // | { type: 'squish'; item: IR; maxWidth: number }
    | { type: 'inline'; items: IR[]; style?: Style }
    | { type: 'horiz'; items: IR[]; style?: Style; wrap?: number } // number indicates indent amount
    | { type: 'indent'; item: IR; amount?: number; style?: Style }
    | { type: 'switch'; options: IR[] }
    | { type: 'loc'; loc: number }
    | { type: 'punct'; text: string; style?: Style };

export type Layout =
    | { type: 'horiz'; wrap?: number }
    | {
          type: 'vert';
          layout: {
              tightFirst: number;
              // This is funny, because it's a list of the LHS locs that shouldn't be squished.
              pairs?: number[];
              indent: number;
          };
      }
    | { type: 'pairs' }
    | { type: 'switch' };

export const nodeToIR = (
    node: Node,
    styles: Record<number, Format>,
    layouts: Record<number, Layout>,
    names: Record<string, Record<number, string>>,
): IR => {
    switch (node.type) {
        case 'array':
        case 'list':
        case 'record': {
            const lr = {
                array: '[]',
                list: '()',
                record: '{}',
            }[node.type];

            const l = layouts[node.loc] ?? { type: 'switch' };

            const items = node.items.map((loc): IR => ({ type: 'loc', loc }));

            switch (l.type) {
                case 'switch':
                    return {
                        type: 'horiz',
                        items: [
                            { type: 'punct', text: lr[0] },
                            {
                                type: 'switch',
                                options: [
                                    { type: 'horiz', items },
                                    {
                                        type: 'vert',
                                        items,
                                        layout: { tightFirst: 1, indent: 2 },
                                    },
                                ],
                            },
                            { type: 'punct', text: lr[1] },
                        ],
                    };
                case 'horiz':
                    return {
                        type: 'horiz',
                        wrap: l.wrap,
                        items: [
                            { type: 'punct', text: lr[0] },
                            ...items,
                            { type: 'punct', text: lr[1] },
                        ],
                    };
                case 'vert':
                    return {
                        type: 'horiz',
                        items: [
                            { type: 'punct', text: lr[0] },
                            { type: 'vert', items: items, layout: l.layout },
                            { type: 'punct', text: lr[1] },
                        ],
                    };
            }
            throw new Error('unknown layout');
        }

        // case 'rich-text':
        //     return { type: 'text', text: 'rich' };

        case 'id':
            return {
                type: 'text',
                text: node.ref
                    ? names[node.ref.toplevel][node.ref.loc]
                    : node.text,
                style: styles[node.loc] ?? node.ref ? refStyle : undefined,
            };

        case 'string': {
            const multi =
                node.first.includes('\n') ||
                node.templates.some((t) => t.suffix.includes('\n'));
            return {
                type: 'horiz',
                items: [
                    { type: 'loc', loc: node.tag },
                    { type: 'punct', text: '"' },
                    {
                        type: multi ? 'inline' : 'horiz',
                        style: multi
                            ? {
                                  background: 'red',
                              }
                            : undefined,
                        items: [
                            { type: 'text', text: node.first },
                            ...node.templates.flatMap((t): IR[] => [
                                { type: 'punct', text: '${' },
                                { type: 'loc', loc: t.expr },
                                { type: 'punct', text: '}' },
                                { type: 'text', text: t.suffix },
                            ]),
                        ],
                    },
                    { type: 'punct', text: '"' },
                ],
            };
        }

        case 'record-access':
            return {
                type: 'horiz',
                wrap: 2,
                items: [
                    { type: 'loc', loc: node.target },
                    ...node.items.map(
                        (loc): IR => ({
                            type: 'horiz',
                            items: [
                                { type: 'punct', text: '.' },
                                { type: 'loc', loc },
                            ],
                        }),
                    ),
                ],
            };
        case 'annot':
            return {
                type: 'horiz',
                items: [
                    { type: 'loc', loc: node.contents },
                    { type: 'punct', text: ':' },
                    { type: 'loc', loc: node.annot },
                ],
            };
        case 'comment':
        case 'spread':
            return {
                type: 'horiz',
                items: [
                    {
                        type: 'punct',
                        text: node.type === 'spread' ? '..' : ';',
                    },
                    { type: 'loc', loc: node.contents },
                ],
            };
    }
};
