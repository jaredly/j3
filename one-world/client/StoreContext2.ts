import { createContext, RefObject, useContext } from 'react';
import { DocSession, PersistedState } from '../shared/state2';
import { Action } from '../shared/action2';
import { Path } from '../shared/nodes';

export type StoreEvt = 'selection' | 'all';

export type Store = {
    session: string;
    docSession: DocSession;
    // getDocSession(doc: string, session?: string): DocSession;
    getState(): PersistedState;
    update(...actions: Action[]): void;
    onSelection(session: string, path: Path, f: () => void): () => void;
    onTop(id: string, f: () => void): () => void;
    onTopNode(top: string, id: number, f: () => void): () => void;
    onDoc(id: string, f: () => void): () => void;
    onDocNode(doc: string, id: number, f: () => void): () => void;
    on(evt: 'selection', f: (autocomplete?: boolean) => void): () => void;
    on(evt: 'all', f: () => void): () => void;
    // focus and drag management
    textRef(path: Path, pathKey: string): RefObject<HTMLElement>;
    startDrag(pathKey: string, path: Path): void;
    dispose(): void;
};

export const StoreContext = createContext<Store>(null as any);

export const useStore = () => {
    return useContext(StoreContext);
};

// export const SessionContext = createContext<string>(null as any);
// export const useSessionId = () => {
//     return useContext(SessionContext);
// };
