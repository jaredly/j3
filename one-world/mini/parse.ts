import { childNodes, Loc, RecNode } from '../shared/nodes';

const findMacros_ = (node: RecNode, found: Loc[]) => {
    if (
        node.type === 'id' &&
        node.ref?.type === 'toplevel' &&
        node.ref.kind === 'macro'
    ) {
        found.push(node.loc);
    }
    if (
        node.type === 'list' &&
        node.items[0].type === 'id' &&
        node.items[0].text === '`'
    ) {
        return; // quoting, abort
    }
    childNodes(node).forEach((child) => findMacros_(child, found));
};

export const findMacros = (node: RecNode) => {
    const found: Loc[] = [];
    findMacros_(node, found);
    return found;
};

type Pattern =
    | { type: 'id'; text: string; loc: Loc }
    | { type: 'constructor'; name: string; args: string[]; loc: Loc }
    | Prim;

type Prim =
    | { type: 'bool'; value: boolean; loc: Loc }
    | { type: 'int'; value: number; loc: Loc };

type Expr =
    | Prim
    | { type: 'var'; text: string; loc: Loc }
    | { type: 'ref'; ref: Loc; loc: Loc }
    | { type: 'function'; args: { name: string; loc: Loc }[]; loc: Loc }
    | { type: 'apply'; target: Expr; args: Expr[]; loc: Loc }
    | {
          type: 'string';
          tag: Expr | null;
          first: string;
          interps: { expr: Expr; suffix: string }[];
          loc: Loc;
      }
    | {
          type: 'if-let';
          table: { pattern: Pattern; value: Expr }[];
          yes: Expr;
          no: Expr;
      }
    | { type: 'quote'; node: RecNode };

/*
match -> if-let
let -> function/apply
if -> if-let
fn -> function

*/

const parseExpr = (node: RecNode): Expr | void => {
    // ok
};

export const parse = (node: RecNode, cursor?: number) => {
    // hrm
};
