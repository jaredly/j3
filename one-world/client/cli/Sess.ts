import { existsSync, readFileSync, writeFileSync } from 'fs';
import { IRSelection } from '../../shared/IR/intermediate';
import { Store } from '../StoreContext2';

const sessFile = './.cli.sess';
export type Sess = { ssid: string; doc?: string; selection?: IRSelection[] };
export const readSess = (): Sess => {
    if (existsSync(sessFile)) {
        return JSON.parse(readFileSync('./.cli.sess', 'utf-8'));
    }
    return { ssid: 'cli' };
};
export const writeSess = (sess: Sess) =>
    writeFileSync(sessFile, JSON.stringify(sess));
export function trackSelection(store: Store, sess: Sess, docId: string) {
    const unsel = store.on('selection', () => {
        sess.selection = store.getDocSession(docId).selections;
        writeSess(sess);
    });

    process.on('beforeExit', () => {
        unsel();
        store.update({ type: 'selection', doc: docId, selections: [] });
    });
    return unsel;
}
