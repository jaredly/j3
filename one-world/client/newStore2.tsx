import { AnyEvaluator } from '../evaluators/boot-ex/types';
import { ClientMessage, ServerMessage } from '../server/run';
import { Action } from '../shared/action2';
import {
    IRCursor,
    IRSelection,
    updateNodeText,
} from '../shared/IR/intermediate';
import { lastChild } from '../shared/IR/nav';
import { Path, serializePath } from '../shared/nodes';
import {
    DocSelection,
    DocSession,
    getTop,
    HistoryItem,
    PersistedState,
} from '../shared/state2';
import {
    applyChanges,
    reverseChanges,
    update,
    Updated,
} from '../shared/update2';
import { getAutoComplete } from './cli/getAutoComplete';
import { RState } from './cli/render';
import { IncomingMessage, OutgoingMessage } from './cli/worker';
import { listen } from './listen';
import { getTopForPath } from './selectNode';
import { Store } from './StoreContext2';

type Evts = {
    general: {
        selection: ((autocomplete?: boolean) => void)[];
        all: (() => void)[];
    };
    selections: Record<string, (() => void)[]>;
    tops: Record<
        string,
        {
            fns: (() => void)[];
            nodes: Record<number, (() => void)[]>;
        }
    >;
    docs: Record<
        string,
        {
            fns: (() => void)[];
            nodes: Record<number, (() => void)[]>;
        }
    >;
};

const blankEvts = (): Evts => ({
    tops: {},
    docs: {},
    selections: {},
    general: { selection: [], all: [] },
});
const blankFns = (): Evts['docs'][''] => ({ fns: [], nodes: {} });

export const ensure = <K extends string | number, A>(
    obj: Record<K, A>,
    k: K,
    n: () => A,
) => {
    if (!obj[k]) {
        obj[k] = n();
    }
    return obj[k];
};

const selPathKeys = (sel: IRSelection) => {
    return sel.end ? [sel.start.key, sel.end.key] : [sel.start.key];
};

const setupDragger = (store: Store) => {
    let dragState = null as null | {
        pathKey: string;
        path: Path;
        node: HTMLElement;
    };
    const textRefs: Record<
        string,
        { path: Path; current: null | HTMLElement }
    > = {};

    const mouseUp = () => {
        document.removeEventListener('mouseup', mouseUp);
        document.removeEventListener('mousemove', mouseMove);
    };

    const mouseMove = (evt: MouseEvent) => {
        if (!dragState) return;

        const sels = store.docSession.selections;
        const matching = sels.find(
            (s) =>
                s.type === 'ir' && !s.end && s.start.key === dragState!.pathKey,
        );

        const range = new Range();
        // const sel = getNewSelection(
        //     matching && !matching?.end
        //         ? { start: matching.start, sel: matching.cursor }
        //         : null,
        //     dragState.node,
        //     { x: evt.clientX, y: evt.clientY },
        //     true,
        //     range,
        // );
        // if (!sel) return;
        // store.update({
        //     type: 'in-session',
        //     action: { type: 'multi', actions: [] },
        //     doc: dragState.path.root.doc,
        //     selections: [
        //         {
        //             type: 'id',
        //             cursor: sel.sel,
        //             start: sel.start,
        //             path: dragState.path,
        //             pathKey: dragState.pathKey,
        //             text: matching?.type === 'id' ? matching.text : undefined,
        //         },
        //     ],
        // });
    };

    const startDrag = (pathKey: string, path: Path) => {
        if (!textRefs[pathKey]) {
            console.warn('No registered', textRefs);
            return;
        }
        if (!textRefs[pathKey].current) {
            console.warn('no what');
        }
        document.addEventListener('mouseup', mouseUp);
        document.addEventListener('mousemove', mouseMove);
        dragState = { pathKey, path, node: textRefs[pathKey].current! };
    };

    return { textRefs, startDrag };
};

export function recalcDropdown(
    store: Store,
    docId: string,
    rstate: RState,
    ev: AnyEvaluator,
) {
    const ds = store.docSession;
    if (!ds.dropdown || ds.dropdown.dismissed) {
        const auto = getAutoComplete(store, rstate, ds, ev);
        if (auto?.length) {
            ds.dropdown = { selection: [0] };
        } else {
            ds.dropdown = undefined;
        }
    }
}

export type WS = {
    send(msg: ClientMessage): void;
    onMessage(fn: (msg: ServerMessage) => void): void;
    close(): void;
};

export const newStore = (
    state: PersistedState,
    ws: WS,
    session: DocSession | null,
): Store => {
    const evts = blankEvts();
    // const docSessionCache: { [id: string]: DocSession } = {};
    // TODO have a way to identify other users, name, pic, etc.
    const presence: Record<string, DocSelection[]> = {};

    if (!session) {
        session = {
            doc: state.id,
            jumpHistory: [],
            selections: [],
            clipboard: [],
        };
    }

    const store: Store = {
        session: 'nop',
        dispose() {
            ws.close();
            evts.docs = {};
            evts.general = { all: [], selection: [] };
            evts.selections = {};
            evts.tops = {};
        },
        docSession: session,
        getState() {
            return state;
        },
        update(...actions) {
            // console.log('store.update', actions.length);
            const prevState = state;

            const updated: Updated = { toplevels: {}, selections: {} };

            const extras: Action[] = [];
            let changes: HistoryItem['changes'] = {
                prevSelections: this.docSession.selections,
                selections: [],
            };

            let presenceChanged = false;

            actions.forEach((action) => {
                if (action.type === 'drag') {
                    const ds = this.docSession;
                    ds.dragState = action.drag;
                    evts.general.selection.forEach((f) => f());
                }

                if (action.type === 'selection') {
                    const prev = session.selections;
                    session.selections = action.selections;
                    session.verticalLodeStone = action.verticalLodeStone;
                    const seen: Record<string, IRSelection> = {};
                    action.selections.forEach((sel) => {
                        if (sel.type !== 'ir') return;
                        selPathKeys(sel).forEach((k) => {
                            const id = `${session}#${k}`;
                            seen[id] = sel;
                            updated.selections[id] = true;
                        });
                    });

                    prev.forEach((psel) => {
                        if (psel.type !== 'ir') {
                            // If you don't manually "commit" a namespace change,
                            // we just reset it.
                            return;
                        }
                        selPathKeys(psel).forEach((k) => {
                            const id = `${session}#${k}`;
                            updated.selections[id] = true;
                            if (
                                psel.start.cursor.type === 'text' &&
                                psel.start.cursor.end.text
                            ) {
                                const cur = seen[id];
                                if (
                                    !cur ||
                                    cur.start.cursor.type !== 'text' ||
                                    cur.start.cursor.end.index !==
                                        psel.start.cursor.end.index
                                ) {
                                    // Only commit if we haven't done an update on that node in this call
                                    // to .update
                                    if (
                                        !updated.toplevels[
                                            psel.start.path.root.toplevel
                                        ]?.[lastChild(psel.start.path)]
                                    ) {
                                        maybeCommitTextChange(
                                            psel.start.cursor,
                                            psel.start.path,
                                            state,
                                            extras,
                                        );
                                    }
                                }
                            }
                        });
                    });

                    evts.general.selection.forEach((f) =>
                        f(action.autocomplete),
                    );

                    if (!action.autocomplete) {
                        session.dropdown = undefined;
                    }
                }

                if (action.type === 'presence') {
                    presence[action.id] = action.selections;
                    presenceChanged = true;
                    // Don't send to websocket!
                    return;
                }

                if (action.type === 'undo') {
                    const last = state.history[state.history.length - 1];
                    if (!last) return state;
                    const rev = reverseChanges(last.changes);
                    state = applyChanges(state, rev);
                    state = {
                        ...state,
                        history: [
                            ...state.history,
                            {
                                changes: rev,
                                idx: state.history.length,
                                reverts: last.idx,
                                session: this.session,
                            },
                        ],
                    };
                    this.docSession.selections = rev.selections;
                    return;
                }

                state = update(state, action, updated, changes);
            });

            ({ changes, state } = maybeAddHistoryItem(
                state,
                changes,
                this.docSession.selections,
                this.session,
            ));

            extras.forEach((action) => {
                state = update(state, action, updated, changes);
            });

            ({ changes, state } = maybeAddHistoryItem(
                state,
                changes,
                this.docSession.selections,
                this.session,
            ));

            evts.general.all.forEach((f) => f());

            const added = state.history.slice(prevState.history.length);
            if (added.length) {
                ws.send({ type: 'changes', items: added });
            }

            // @ts-ignore
            self.state = state;
            // TODO:if (presenceChanged) evts.general.presence.forEach(f => f());
            sendUpdates(updated, evts);
        },
        on(evt, f) {
            switch (evt) {
                case 'all':
                    return listen(evts.general.all, f);
                case 'selection':
                    return listen(evts.general.selection, f);
            }
        },
        onSelection(session, path, f) {
            const id = `${session}#${serializePath(path)}`;
            ensure(evts.selections, id, () => []);
            return listen(evts.selections[id], f);
        },
        onTop(id, f) {
            ensure(evts.tops, id, blankFns);
            return listen(evts.tops[id].fns, f);
        },
        onTopNode(top, id, f) {
            ensure(evts.tops, top, blankFns);
            ensure(evts.tops[top].nodes, id, () => []);
            return listen(evts.tops[top].nodes[id], f);
        },
        onDoc(id, f) {
            ensure(evts.docs, id, blankFns);
            return listen(evts.docs[id].fns, f);
        },
        onDocNode(doc, id, f) {
            ensure(evts.docs, doc, blankFns);
            ensure(evts.docs[doc].nodes, id, () => []);
            return listen(evts.docs[doc].nodes[id], f);
        },

        // Drag stuff
        textRef(path, pathKey) {
            if (!dragger.textRefs[pathKey]) {
                dragger.textRefs[pathKey] = { current: null, path };
            }
            return dragger.textRefs[pathKey];
        },
        startDrag(pathKey, path) {
            dragger.startDrag(pathKey, path);
        },
    };

    ws.onMessage((data) => {
        switch (data.type) {
            case 'presence':
                store.update(data);
                return;
            case 'changes':
                if (data.items[0].idx !== state.history.length) {
                    throw new Error(`history out of sync`);
                }
                state = { ...state, history: state.history.concat(data.items) };
                for (let item of data.items) {
                    state = applyChanges(state, item.changes);
                }
                // TOOD process
                return;
        }
    });

    const dragger = setupDragger(store);

    return store;
};

const maybeAddHistoryItem = (
    state: PersistedState,
    changes: HistoryItem['changes'],
    selections: DocSelection[],
    session: string,
) => {
    if (changes.nodes || changes.toplevels) {
        // console.log('have some changes');
        changes.selections = selections;
        const historyItem: HistoryItem = {
            idx: state.history.length,
            changes,
            session,
        };
        state = { ...state, history: [...state.history, historyItem] };

        changes = {
            prevSelections: selections,
            selections: [],
        };
    }
    return { state, changes };
};

function sendUpdates(updated: Updated, evts: Evts) {
    Object.entries(updated.toplevels).forEach(([top, nodes]) => {
        Object.keys(nodes).forEach((k) => {
            evts.tops[top]?.nodes[+k]?.forEach((f) => f());
        });
        evts.tops[top]?.fns.forEach((f) => f());
    });

    Object.keys(updated.selections).forEach((id) =>
        evts.selections[id]?.forEach((f) => f()),
    );
}

function maybeCommitTextChange(
    sel: Extract<IRCursor, { type: 'text' }>,
    path: Path,
    state: PersistedState,
    extras: Action[],
) {
    const last = path.children[path.children.length - 1];
    const node = getTopForPath(path, state)?.nodes[last];

    if (!sel.end.text || !node) return;
    const updated = updateNodeText(node, sel.end.index, sel.end.text.join(''));
    if (!updated) {
        return;
    }

    // console.log('committing text change');

    extras.push({
        type: 'toplevel',
        id: path.root.toplevel,
        doc: path.root.doc,
        action: {
            type: 'update',
            update: { nodes: { [node.loc]: updated } },
        },
    });
}
