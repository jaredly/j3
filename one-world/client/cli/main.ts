import { drop } from './edit/drop';
import { handleMouseClick, handleMouseDrag } from './edit/handleMouse';
import {
    handleDropdown,
    handleMovement,
    handleUpDown,
} from './edit/handleMovement';
import { handleUpdate } from './edit/handleUpdate';
// import { genId, newDocument } from './edit/newDocument';
import { handleDrag, maybeStartDragging } from './handleDrag';
import { pickDocument } from './pickDocument';
import { parseAndCache, render, renderSelection } from './render';
import { Sess, trackSelection } from './Sess';

import { SimplestEvaluator } from '../../evaluators/simplest';
import { Caches, Context } from '../../graphh/by-hand';
import { ABlock, toChunk } from '../../shared/IR/block-to-attributed-text';
import { recalcDropdown } from '../newStore2';
import { Store } from '../StoreContext2';
import {
    drawToTerminal,
    MouseEvt,
    MouseKind,
    Renderer,
} from './drawToTerminal';
import { OutgoingMessage } from './worker';
import { previewDocument } from './previewDocument';
import { timeoutTracker } from './timeoutTracker';
import { genId } from './edit/newDocument';

export const run = async (term: Renderer) => {
    console.log('initializing store...');
    let sess = term.readSess();

    if (!sess) {
        // TODO:
        const docs = await term.docList();
        const picked = await pickDocument(docs, term, (id) => {});
        console.log('picked', picked);
        if (picked.id === null) {
            const store = await term.newDoc(picked.title);
            return runDocument(term, store, {
                doc: store.getState().id,
                selection: [],
                ssid: store.session,
            });
            // OK SO here, I think we ...
            // ... initialize a store around that stage?
            // I think.
            // like, the main idea is that we're shrinking
            // the visible surface area of `persistedstate`
            // and the store generally, to just encompass
            // the `DocStage`, which contains a copy of
            // all relevant toplevels.
        }
        const store = await term.loadDoc(picked.id);
        return runDocument(term, store, {
            doc: picked.id,
            selection: [],
            ssid: genId(),
        });
    }

    const store = await term.loadDoc(sess.doc);
    await runDocument(term, store, sess);

    setTimeout(() => {
        return process.exit(0);
    }, 50);
};

export function runDocument(term: Renderer, store: Store, sess: Sess) {
    console.log('running a doc', store);
    // const store = await term.init(sess);
    const docId = store.getState().id;

    let lastKey = null as null | string;
    const ev = SimplestEvaluator;

    const handleMessage = (msg: OutgoingMessage) => {
        tracker.receivedMessage(msg.id);
        Object.keys(msg.output).forEach((id) => {
            rstate.parseAndEval[id].output = msg.output[id];
        });

        rstate = render(term.width - 10, store, docId, rstate.parseAndEval);
        if (needsDropdownRecalc) {
            recalcDropdown(store, docId, rstate, ev);
            needsDropdownRecalc = false;
        }
        drawToTerminal(rstate, term, store, docId, lastKey, ev);
    };

    let worker = term.spawnWorker(handleMessage);

    const restartWorker = () => {
        worker.terminate();
        worker = term.spawnWorker(handleMessage);
        // Send latest infos
        const { caches, ctx } = parseAndCache(store, docId, {}, ev);
        sendToWorker(caches, ctx);
    };
    const tracker = timeoutTracker(restartWorker);

    const sendToWorker = (caches: Caches<unknown>, ctx: Context) => {
        worker.sendMessage({
            type: 'evaluates',
            id: tracker.nextMsgId(),
            ctx,
            evid: 'simplest',
            caches,
            tops: Object.keys(caches.parse),
        });
    };

    let resolve = (quit: boolean) => {};
    const finisher = new Promise<boolean>((res) => (resolve = res));

    const cleanup: (() => void)[] = [
        () => store.update({ type: 'selection', doc: docId, selections: [] }),
    ];
    const clean = (v: () => void) => cleanup.push(v);

    const finish = (quit: boolean) => {
        cleanup.forEach((v) => v());
        setTimeout(() => {
            resolve(quit);
        }, 50);
    };

    // The parse
    const { parseCache, caches, ctx } = parseAndCache(store, docId, {}, ev);
    let rstate = render(term.width - 10, store, docId, parseCache);
    drawToTerminal(rstate, term, store, docId, lastKey, ev);
    clean(trackSelection(store, sess, docId, term.writeSess));
    sendToWorker(caches, ctx);

    let prevState = store.getState();
    let tid: null | Timer = null;
    let needsDropdownRecalc = false;

    const rerender = () => {
        tid = null;

        // The parse
        const { parseCache, caches, ctx } = parseAndCache(
            store,
            docId,
            rstate.parseAndEval,
            ev,
        );

        // The eval now happens in the worker

        rstate = render(term.width - 10, store, docId, parseCache);
        if (needsDropdownRecalc) {
            recalcDropdown(store, docId, rstate, ev);
            needsDropdownRecalc = false;
        }
        drawToTerminal(rstate, term, store, docId, lastKey, ev);
        prevState = store.getState();

        sendToWorker(caches, ctx);
    };

    const kick = () => {
        if (tid != null) return;
        tid = setTimeout(rerender, 0);
    };

    clean(
        store.on('selection', (autocomplete) => {
            if (autocomplete) {
                needsDropdownRecalc = true;
                kick();
            }
        }),
    );

    clean(
        store.on('all', () => {
            kick();
        }),
    );

    clean(term.onResize(() => kick()));

    clean(
        term.onKey((key: string) => {
            if (onKey(key)) {
                return;
            }

            if (key === 'ESCAPE') {
                return finish(true);
            }

            term.moveTo(1, term.height - 1, [
                [toChunk(key, { background: { r: 0, g: 0, b: 0 } })],
            ]);
            renderSelection(term, store, docId, rstate.sourceMaps);
        }),
    );

    clean(
        term.onMouse((evtKind: MouseKind, evt: MouseEvt) => {
            onMouse(evtKind, evt);
        }),
    );

    const onKey = (key: string) => {
        lastKey = key;
        if (key === 'CTRL_W') {
            finish(false);
            return true;
        }

        if (handleDropdown(key, docId, store, rstate, kick, ev)) {
            return true;
        }

        if (
            handleUpDown(key, docId, store, rstate) ||
            handleMovement(key, docId, rstate.cache, store)
        ) {
            return true;
        }
        if (handleUpdate(key, docId, rstate.cache, store)) {
            return true;
        }
    };

    const onMouse = (evtKind: MouseKind, evt: MouseEvt) => {
        const ds = store.getDocSession(docId);
        if (evtKind === 'MOUSE_DRAG') {
            if (ds.dragState) {
                handleDrag(evt, docId, rstate, ds.dragState, store);
                return;
            }
            handleMouseDrag(
                docId,
                rstate.sourceMaps,
                evt,
                store,
                rstate.dropTargets,
                rstate.cache,
            );
        } else if (evtKind === 'MOUSE_LEFT_BUTTON_PRESSED') {
            const sel = ds.selections[0];
            if (
                sel?.type === 'ir' &&
                ds.selections.length &&
                maybeStartDragging(evt, docId, rstate, sel, store)
            ) {
                return;
            }

            // if (evt.y >= term.height - 1) {
            // }
            handleMouseClick(
                docId,
                rstate.sourceMaps,
                rstate.dropTargets,
                evt,
                rstate.cache,
                store,
            );
        } else if (evtKind === 'MOUSE_LEFT_BUTTON_RELEASED') {
            const ds = store.getDocSession(docId);
            const dragState = ds.dragState;
            if (dragState) {
                store.update({ type: 'drag', doc: docId, drag: undefined });
                if (!dragState.dest) {
                    handleMouseClick(
                        docId,
                        rstate.sourceMaps,
                        rstate.dropTargets,
                        evt,
                        rstate.cache,
                        store,
                    );
                } else {
                    drop(dragState.source, dragState.dest, store);
                }
            }

            return;
        }
    };

    return finisher;
}
