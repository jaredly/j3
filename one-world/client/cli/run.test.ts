import { test, expect } from 'bun:test';
import { run } from './main';
import {
    KeyFn,
    MouseEvt,
    MouseFn,
    MouseKind,
    Renderer,
} from './drawToTerminal';
import { newStore } from '../newStore2';
import { ABlock } from '../../shared/IR/block-to-attributed-text';
import { splitGraphemes } from '../../../src/parse/splitGraphemes';
import { Sess } from './Sess';
import { DocSession, PersistedState } from '../../shared/state2';
import { Store } from '../StoreContext2';

type Buffer = {
    rows: string[][];
    width: number;
    pos: { x: number; y: number };
    prints: { block: ABlock; pos: { x: number; y: number } }[];
};

const init = (): Buffer => ({
    rows: [[]],
    width: 0,
    pos: { x: 0, y: 0 },
    prints: [],
});
const moveTo = (buf: Buffer, x: number, y: number) => {
    buf.pos = { x, y };
};
const setWidth = (buf: Buffer, width: number) => {
    if (width <= buf.width) return;
    const extra: string[] = [];
    for (let i = width; i < buf.width; i++) {
        extra.push(' ');
    }
    buf.rows.forEach((row) => {
        row.push(...extra);
    });
};
const setHeight = (buf: Buffer, height: number) => {
    if (height <= buf.rows.length) return;
    const extra: string[] = [];
    for (let i = 0; i < buf.width; i++) {
        extra.push(' ');
    }
    for (let i = buf.rows.length; i < height; i++) {
        buf.rows.push(extra.slice());
    }
};

const write = (buf: Buffer, text: ABlock) => {
    buf.prints.push({ block: text, pos: { ...buf.pos } });
    const lines = text.map((line) => {
        let text: string[] = [];
        // let len = 0
        line.forEach((chunk) => {
            text.push(...splitGraphemes(chunk.text));
            // len += chunk.len
        });
        return text;
    });
    const maxWidth = lines.reduce((m, l) => Math.max(m, l.length), 0);
    setWidth(buf, buf.pos.x + maxWidth);
    setHeight(buf, buf.pos.y + lines.length);
    for (let y = 0; y < lines.length; y++) {
        const ri = buf.pos.y + y;
        const line = lines[y];
        buf.rows[ri].splice(buf.pos.x, line.length, ...line);
    }
    buf.pos.y += lines.length - 1;
    buf.pos.x += lines[lines.length - 1].length;
};

const print = (buf: Buffer) =>
    buf.rows.map((row) => row.join('').trimEnd()).join('\n');
// '\n' +
// JSON.stringify(buf.prints, null, 2);

const testRenderer = (
    { width, height }: { width: number; height: number },
    state?: PersistedState,
    docSess?: DocSession,
) => {
    const keys: KeyFn[] = [];
    const mouse: MouseFn[] = [];
    const buf = init();
    let sess: Sess = { ssid: 'test-sess' };
    if (docSess && state) {
        if (!state._documents[docSess.doc]) {
            throw new Error(`missing doc ${docSess.doc} in state`);
        }
        sess.doc = docSess.doc;
    }

    let store: Store | null = null;

    const term: Renderer = {
        width,
        height,
        drawCursor() {
            // noop
        },
        moveTo(x, y, text) {
            moveTo(buf, x, y);
            if (text) {
                write(buf, text);
            }
        },
        clear() {
            buf.rows = [[]];
            buf.width = 0;
        },
        async init(sess) {
            // hrm
            store = newStore(
                state ?? {
                    _toplevels: {},
                    _documents: {},
                    modules: {},
                    stages: {},
                },
                {
                    onMessage(fn) {
                        //
                    },
                    send(msg) {
                        //
                    },
                    close() {},
                },
                'test-sess',
                (id) => docSess ?? null,
            );
            return store;
        },
        onKey(fn) {
            keys.push(fn);
            return () => keys.splice(keys.indexOf(fn), 1);
        },
        onMouse(fn) {
            mouse.push(fn);
            return () => mouse.splice(mouse.indexOf(fn), 1);
        },
        onResize: () => () => {},
        readSess() {
            return sess;
        },
        spawnWorker(onMessage) {
            return {
                sendMessage(msg) {
                    //
                },
                terminate() {
                    //
                },
            };
        },
        write(text) {
            write(buf, text);
        },
        writeSess(s) {
            sess = s;
        },
    };
    return {
        term,
        print: () => print(buf),
        key(key: string) {
            keys.forEach((fn) => fn(key));
        },
        mouse(which: MouseKind, evt: MouseEvt) {
            mouse.forEach((fn) => fn(which, evt));
        },
        getStore() {
            return store;
        },
    };
};

const wait = (n: number) => new Promise((res) => setTimeout(res, n));

test('basic empty', async () => {
    const renderer = testRenderer({ width: 200, height: 5 });
    run(renderer.term);
    await wait(10);
    expect(renderer.print()).toEqual('\nNew Document');
});

test('basic empty, make doc', async () => {
    const renderer = testRenderer({ width: 200, height: 3 });
    run(renderer.term);
    await wait(10);
    renderer.key('ENTER');
    await wait(100);
    expect(renderer.print()).toMatchSnapshot();
});
