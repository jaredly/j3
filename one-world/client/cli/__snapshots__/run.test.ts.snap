// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`basic empty, make doc 1`] = `
"
▶️

Session 10/2/2024, 2:22:05 PM"
`;
